### ISSUE TITLE
[Request] SteamWorld Heist - FPS cap

### ISSUE THREAD

**Issue by BFG1992** [Electry/VitaGrafixPatchlist#64]
SteamWorld Heist is a charming and incredibly fun indie turn-based tactical game, with dark steampunky environments which look especially gorgeous on that phat Vita's OLED screen. The only problem is that, for some weird reason it is locked to 30 FPS. Why weird? Well, that's because it runs fullspeed _even under LOLIcon's battery saving mode_ (yes, that's the same mode where even the LiveArea slows down and the absolute most games just become slideshow with stuttery sound), slowing down only during explosions - yeah, it's optimized THAT well (kudos to Image & Form!).
While the game is well playable in 30 FPS, I (and many other Heist players on Vita, I think) would really appreciate if someone could make a FPS uncapping patch to patchlist, because there's just no damn reason why this game can't hit 60 FPS on Vita, even without an overclock.

_P.S.: Sorry for closing-reopening and edits. That was result of me accidentally pressing Enter instead of another key, realizing in horror that it got submitted without the ability to delete, then editing it in panic :(_

**Comment by q6164007** [comment]
```
# SteamWorld Heist  [EU 1.02]
[PCSB00693,eboot.bin,0xBD7CD4EB]
# SteamWorld Heist  [US 1.02]
[PCSE00583,eboot.bin,0x51AB9D2C]
@FPS
0:0x2AA7F0 nop() *2
```

**Comment by BFG1992** [comment]
@q6164007 Thank you for looking into this! However, the game crashes at launch with error C2-12828-1 if the FPS is set to anything but OFF in config file. Tried both US and EU versions of the game. 

Steps:
1. Copy-pasted your code at the end of patchlist.txt file on my Vita.
2. Set the FPS parameter to 30 or 60 in VitaGrafix Configurator or in the VitaGrafix' config.txt.
3. Launch the game -> crash after the splash screen.

Here's a vid of the issue if you need it: https://youtu.be/W0cn1c4p7xo

**Comment by q6164007** [comment]
Sorry, I don't know why your game crashes. It works well in my Psvita. Maybe you can restart your Psvita and check other plug-ins.
![2019-08-06-070809](https://user-images.githubusercontent.com/48059442/62500806-9fbaf100-b81a-11e9-9fcb-414e97a97e70.png)

**Comment by BFG1992** [comment]
@q6164007 Well, I tried to disable all of my plugins except those necessary for launching the game and the VitaGrafix, namely: VitaGrafix itself (duh), ioplus, NoNpDRM and YAMT (I use it instead of StorageMgr and had no problems at all with it). I also left Adrenaline on, because it shouldn't affect Vita apps anyway (in theory).
Result: crashes as before :(

I have a PCH-1008 Vita on 3.60 Enso, using US version of the game with 1.02 patch and both DLCs installed. Do you have any ideas on what might be causing the crash?

Also, maybe you can share your VitaGrafix's patchlist.txt and config.txt files? I'm thinking that maybe I copy-pasted something wrong so I could use your files... I'm just running out of ideas here, honestly :(

**Comment by q6164007** [comment]
After watching your video, I noticed that the VitaGrafix version you used was 4.1, and you could try upgrading to 4.1.1. You can find a log.txt file at `ux0:/data/VitaGrafix/` and show it. Like this, it can help us analyze the cause of the error.
```
[MAIN] Title ID: PCSE00583
[MAIN] SELF: ux0:/patch/PCSE00583/eboot.bin
[MAIN] NID: 0x51AB9D2C
=======================================
[CONFIG] Parsing config.txt
[CONFIG] Found [MAIN] section, pos=0
[CONFIG] Found [PCSE00583] section, pos=112
[PATCH] Parsing patchlist.txt
[MAIN] Patching seg000 : 002AA7F4 to 00 BF 00 BF, size=4
[PATCH] 262 total patches found in patchlist.txt
```
You can copy the code you need from my patchlist. txt file, but it's better not to use it directly, because there's a lot of my experimental code in it.
[patchlist.txt](https://github.com/Electry/VitaGrafixPatchlist/files/3475164/patchlist.txt)
[config.txt](https://github.com/Electry/VitaGrafixPatchlist/files/3475165/config.txt)

**Comment by BFG1992** [comment]
@q6164007 Sorry for such a late response (3 weeks... damn)! The reason for this is that I have written a long and detailed answer _using my phone_ with how I got it working, and I thought to add an emoji at the end, so I have opened a new tab with GitHub's emoji list... which froze my browser app. I tried to let my phone lay idle for half of an hour in hope that the app will sort its stuff out and unfreeze, letting me to submit my comment, but that didn't happen. I became so demotivated that I decided to re-write the answer tomorrow. And then the next tomorrow. And then once more. Then your comment about logs with linked files got deleted, I'm not sure why that happened. Finally, today I thought that I wanted to thank you anyway (and close the issue too), because I did got it working, and I'm really happy about that! :)

So about the important stuff. Thanks for mentioning the logs! There were 2 lines that looked suspicious:
```
[IO] WARN: Last line is missing EOL char!
[MAIN] Patching seg000 : 002AA7F4 to 00 BF 00 BF 00 BF 00 BF 00 BF 00 BF, size=12
```
So I went to the patchlist.txt with VitaShell and added an empty line after your patch's code, and BAM. Everything works like a charm: the game launches (no crash), the FPS is 60, everything is nice and smooth (except an overclock to 444/222 was needed - guess my point about "no damn reason the game shouldn't be at 60 FPS by default" is not valid". The same log line now looks as in your now removed message:
```
[MAIN] Patching seg000 : 002AA7F4 to 00 BF 00 BF, size=4
```
The issue occurred because I've used an advanced text editor app (with line numbers and stuff, much like Notepad++ on Windows) to paste the code into patchlist.txt, and not added a new line after. I'm pointing at this editor because I later, out of curiosity, deleted the added empty line using VitaShell and the game still launches in 60 FPS without an issue. When viewing the patchlist in that editor then, the file looks with an empty line still. Oh, and after adding an empty line with VitaShell, there's now two empty lines in that editor. Weird :)

So, thank you for your work (and for reading all this, lol)! I really appreciated that :)

**tl;dr:** Read logs, saw weird stuff, added an empty line after the patch's code, everything works.
